### PR TITLE
Fixed issue where the delegation was forced to queries instead of delegating to the proper root type.

### PR DESCRIPTION
**Pull request by OneCyrus**
Mutation stitching paths were forced to the query tree when not a root type.

Addresses #2087

### CODE REVIEW

**Comment by michaelstaib** [comment]
Thanks, @OneCyrus for this fix. There seem to be some snapshots broken.

**Comment by OneCyrus** [comment]
i was already looking into this snapshot issues a bit. but there seems to be some special cases in these tests. they seem to reuse the types from the query tree and rely on the types to be there. not sure if this is really the intended behaviour or if there are edge cases which don't work after this change. may be you have better insights what really should going on there. but most likely the test scenarios are not correct.

File: src/Stitching/Stitching/Middleware/DelegateToRemoteSchemaMiddleware.cs [hunk 1]
```diff
@@ -282,7 +283,12 @@ private static Path RewriteErrorPath(IError error, Path path)
 
             IStitchingContext stitchingContext = context.Service<IStitchingContext>();
             ISchema remoteSchema = stitchingContext.GetRemoteSchema(schemaName);
-            IComplexOutputType type = remoteSchema.GetOperationType(operationType);
+            IComplexOutputType type = remoteSchema.GetOperationType(OperationType.Mutation);
+            if(type is null)
+            {
+                type = remoteSchema.GetOperationType(OperationType.Query);
+            }
```
Comment (OneCyrus) [inline src/Stitching/Stitching/Middleware/DelegateToRemoteSchemaMiddleware.cs:290]: we should get the operation type from the delegate directive here. how do I access that?

File: src/Stitching/Stitching/Middleware/DelegateToRemoteSchemaMiddleware.cs [hunk 2]
```diff
@@ -33,20 +33,22 @@ public async Task InvokeAsync(IMiddlewareContext context)
...
                 NameString schemaName = default;
+                OperationType operationType = OperationType.Query;
                 IReadOnlyCollection<IError>? errors = null;
 
                 foreach (DelegateDirective delegateDirective in context.Field
                     .Directives[DirectiveNames.Delegate]
                     .Select(t => t.ToObject<DelegateDirective>()))
                 {
                     schemaName = delegateDirective.Schema;
+                    operationType = delegateDirective.Operation;
```
Comment (OneCyrus) [inline src/Stitching/Stitching/Middleware/DelegateToRemoteSchemaMiddleware.cs:44]: the auto stitched root types should enable us to infer the default operation type in a better way. not sure yet how I get access to this information.

**Comment by OneCyrus** [comment]
@michaelstaib the issue is that one use case is stitching query requests into mutation reponses. this requires us to have an extended delegate directive which allows to specify the operation type. my changes allow this to work now but we need to have a way to infer this so we don't break existing code.

**Comment by michaelstaib** [comment]
So actually you should bever need to specify the operation type. You either are stitching a root field which should align with the root type or you are stitching in a `Query` everything else would not be spec-compliant since only fields directly on the root type (Mutation) can be mutations everything beneath is considered a query, thus you can only stitch in queries on this level.

**Comment by michaelstaib** [comment]
with this knowledge the following if-clause has to be:

```csharp
OperationType operationType =
                context.Schema.IsRootType(context.ObjectType)
                    ? context.Operation.Operation
                    : OperationType.Query;
````
which solves everything around this.

**Comment by OneCyrus** [comment]
well it's actually spec compliant. the only thing the spec defines is the execution order on the root types

**Comment by michaelstaib** [comment]
The spec defines execution-algorithms and if we, for instance, would allow mixing in mutations where queries are expected this could lead to some serious side-effects on the downstream services. I would be opposed to allowing that since we could not contain the side effects.

**Comment by michaelstaib** [comment]
@OneCyrus, I have talked to @PascalSenn and we will not merge this into the main. We are thinking to allow combining fields in a more complex way but not for the current version. Merging fields into one always go hand in hand with a proper state machine that also is able to pass along results as variables into the next field.

Regarding, nested mutations I do not see a way forward since the current spec explicitly states that mutations are only the fields on the mutation type and only there are side-effects allowed.

> It is expected that the top level fields in a mutation operation perform side‐effects on the underlying data system. Serial execution of the provided mutations ensures against race conditions during these side‐effects.

The spec not only defines field ordering but speaks of side-effects to the underlying data system.

I am closing this PR and we will pick up some functionality around this later in v11 regarding combining fields.

**Comment by michaelstaib** [comment]
also ...

> Because the resolution of fields other than top‐level mutation fields must always be side effect‐free and idempotent, the execution order must not affect the result, and hence the server has the freedom to execute the field entries in whatever order it deems optimal.

https://spec.graphql.org/draft/#sec-Normal-and-Serial-Execution

**Comment by OneCyrus** [comment]
> also ...
> 
> > Because the resolution of fields other than top‐level mutation fields must always be side effect‐free and idempotent, the execution order must not affect the result, and hence the server has the freedom to execute the field entries in whatever order it deems optimal.

obviously it's bad design to have side effects on mutations. even if HC would handle this it would be problematic if more than one user would use a graphql API. that's why we have only side-effects-free mutations and don't chain mutations at all. also we avoid to stitch queries into mutations as we are running async anyway and don't want to have synchronisation contexts across microservices.

I guess we will fork it for now as we don't have a use case for stitching queries into mutations.
